### PR TITLE
chore(website): add playground to website

### LINKED ISSUE #168: Interactive examples / playground
Spawning off of the discussions in #125, logging it separately for visibility

Would be good if we can setup an interactive playground like https://mysticatea.github.io/vue-eslint-demo/

Similarly it would be good if we can add interactive examples to our docs like eslint-plugin-es: https://mysticatea.github.io/eslint-plugin-es/rules/no-json-superset.html

Both of these would of course require us to generate and deploy some form of a website.

The former would definitely be helpful in having users give us concrete REPLs for their bugs, and the latter is nicer than having to manually annotate static markdown codeblocks.

### PR DESCRIPTION
**Pull request by armano2**
## Overview

Migrated version of playground from #3147 

Fixes #168

- [array-type: default](https://deploy-preview-4108--typescript-eslint.netlify.app/play#ts=4.4.4&sourceType=module&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AIYoqlRbTwgBcATAL5A&code=MYewdgzgLgBAHgLhgQQE6oIYE8A81UCWYA5gHwwC8MA2gLoCwAUEA)
- [array-type: generic](https://deploy-preview-4108--typescript-eslint.netlify.app/play/#ts=4.4.4&jsx=true&sourceType=module&showAST=true&rules=N4IgAgLgngDgpgZwMYCcCWMIFpEBs0B2EA9AIYoqlRbTwgBcA2gEwA0oAJnAGakCuuCAxABzOATjokIAL4BdGUA&code=MYewdgzgLgBAHgLhtATgSzAcwNoF0YC8M2A5AIYkA0MJARibgNwCwAUKJLAJ5IoCmZACbgANl2RR0WPIWLkqNekzZA)

#### TODO:
  - [x] make it work with latest versions of packages
  - [x] add way to load config instead of drop-down with rules
  - [x] add button to generate markdown to copy/paste in tickets
  - [x] in options add dropdown to pick typescript version
  - [x] AST viewer should expand childs when needed
  - [x] add more tsconfig options
  - [x] use lzstringto compress code for url

### CODE REVIEW

File: packages/scope-manager/src/referencer/PatternVisitor.ts
Comment (armano2) [inline packages/scope-manager/src/referencer/PatternVisitor.ts:2]: this change should be ported to master, we have circular dependencies because of that

File: packages/eslint-plugin/src/util/collectUnusedVariables.ts
Comment (armano2) [inline packages/eslint-plugin/src/util/collectUnusedVariables.ts:9]: circular dependencies

**Comment by armano2** [comment]
@JoshuaKGoldberg @bradzacher what do you guys think about building linter for web as separate package, and exposing it as `iife` instead of cjs as this should simplify / optimize loading of linter for website.

--------------

Playground is now working locally, I'm not sure yet why its failing on server,

--------------

we should change build configuration of netlify to use newer version of node

image

**Comment by armano2** [comment]
Small update, simple editor for tsconfig and eslint added

**Comment by armano2** [comment]
What do you guys think about this new config editor:

## Eslint config
image image

## TypeScript Config
image
image

File: packages/website/src/components/Playground.tsx
Comment (JoshuaKGoldberg) [inline packages/website/src/components/Playground.tsx:21]: This is a lot of files and is pretty fully featured. I'll do my best to review (and recognize having the non-MVP features in is useful for seeing where things are going)... but it would be easier if features like copying markdown, showing ASTs, and modifying configs as more than just JSON were split into separate PRs.

Not a blocker, and still shy of a request 😄

File: packages/website/src/components/editor/Editor.tsx
Comment (JoshuaKGoldberg) [inline packages/website/src/components/editor/Editor.tsx:56]: I've had good experiences with https://github.com/react-monaco-editor/react-monaco-editor in the past. Any reason do recreate it here?

Comment (JoshuaKGoldberg) [inline packages/website/src/components/editor/Editor.tsx:56]: I've had good experiences with https://github.com/react-monaco-editor/react-monaco-editor in the past. Any reason not to use it here?

Comment (armano2) [reply packages/website/src/components/editor/Editor.tsx:56]: we are actually not using monaco, we are using sandbox env from https://www.typescriptlang.org/dev/sandbox/

File: packages/website/src/components/Playground.tsx
Comment (armano2) [reply packages/website/src/components/Playground.tsx:21]: i'm actually not sure about config formats, currently i'm accepting reading `compilerOptions` and `rules` from config

File: packages/website/src/components/editor/Editor.tsx
Comment (armano2) [reply packages/website/src/components/editor/Editor.tsx:46]: i did few improvements to editor, we have new api now in sandbox and i was able to remove some of code

File: packages/eslint-plugin/src/util/getWrappingFixer.ts
Comment (armano2) [inline packages/eslint-plugin/src/util/getWrappingFixer.ts:7]: i missed this in #4140

File: packages/website/src/components/hooks/useFocus.ts
Comment (JoshuaKGoldberg) [inline packages/website/src/components/hooks/useFocus.ts:3]: Nit: the `T` is never used.

```suggestion
function useFocus(): [
```

File: packages/website/src/components/editor/Editor.tsx
Comment (JoshuaKGoldberg) [inline packages/website/src/components/editor/Editor.tsx:23]: The `Editor` is only called in one place, so all these optional ones don't actually need to be optional. Which should then mean that `if (` statements for them below shouldn't be necessary.

File: packages/website/src/components/editor/Editor.tsx
Comment (JoshuaKGoldberg) [inline packages/website/src/components/editor/Editor.tsx:13]: I think this's be a little easier to read with a `configModel: ConfigModel;` inside the props instead of extending. That way it can take in a prop more like `state={state}` the way `OptionsSelector` does.

File: packages/website/src/components/hooks/useHashState.ts
Comment (JoshuaKGoldberg) [inline packages/website/src/components/hooks/useHashState.ts:147]: Nit: you might consider `react-use`'s `useEvent` here.

**Review by JoshuaKGoldberg — APPROVED** [review]
The playground works great and would be a huge benefit as-is, thank you for (again) pressing on this! ❤️‍🔥 

I don't want to block this going in on refactors, so just leaving thoughts on the PR without blocking.

File: packages/website/src/components/hooks/useHashState.ts
Comment (armano2) [reply packages/website/src/components/hooks/useHashState.ts:68]: you are correct, this is no longer needed as i'm no longer rendering this page on server

File: packages/website/src/components/lib/debounce.ts
Comment (armano2) [reply packages/website/src/components/lib/debounce.ts:1]: wdym, `Lodash` for web? we should not have it

File: packages/website-eslint/package.json
Comment (armano2) [reply packages/website-eslint/package.json:30]: atleast for now, build is kinda specific for us, i'm including `typescript-eslint` packages and exposing it as `amd` that afterwards is loaded in sandbox

File: packages/website/src/components/editor/Editor.tsx
Comment (armano2) [reply packages/website/src/components/editor/Editor.tsx:83]: i like that idea, i'm going to apply first all changes and after do this

Comment (armano2) [reply packages/website/src/components/editor/Editor.tsx:85]: ok, this is not going to be that easy, can we move refactor after first release?

File: packages/website/src/components/lib/debounce.ts
Comment (JoshuaKGoldberg) [reply packages/website/src/components/lib/debounce.ts:1]: I see `camelcase`, `debounce`, and `merge` in the dev tools file name search:

Screenshot of Chrome dev tools loading camelcase, debounce, and merge from lodash

Comment (armano2) [reply packages/website/src/components/lib/debounce.ts:1]: i see where you are coming from now, code of bundled rules is isolated from web and we do not have direct access to it (that on purpose), to limit potential leaks of functionality or collisions

File: packages/website/src/components/editor/Editor.tsx
Comment (armano2) [reply packages/website/src/components/editor/Editor.tsx]: i looked into your example and i devised something simmilar (havilly inspirated), but your implementation had bunch of issues (i know)

--------------------------

i decided to refactor has state handling for easier time checking what is going on i added some info log events, let me know what do you think

as for updating code: i solved it by doing checks but generally issue lays down in fact that we have 2 dynamic sources of changes

1. hash
2. event from editor

when hash changes editor updates, thus editor emit event than hash updates, and we have loop :P

that was partially solved before by isEditing, but this is not going to work in composite api

(this is hard to explain)

File: packages/website/src/components/hooks/useHashState.ts
Comment (armano2) [reply packages/website/src/components/hooks/useHashState.ts:158]: i don't like libraries that are doing to simple thinks, its overkill and adding it just to replace seem like overkill
```ts
  useEffect(() => {
    window.addEventListener('resize', resize);
    return (): void => {
      window.removeEventListener('resize', resize);
    };
  });
```

File: packages/website/src/components/modals/Modal.tsx
Comment (armano2) [reply packages/website/src/components/modals/Modal.tsx:13]: i don't mind using components, and definitly that could be usefull, i'm just unsure what stack/library we should use 

if we decide to use reach than we should just use

https://reach.tech/accordion - for options
https://reach.tech/checkbox - instead of my implementation
https://reach.tech/tooltip - for tooltips after copying

is this component library maintained, are there better choices? there is lots of questions :P

File: packages/website/src/components/Playground.module.css
Comment (armano2) [reply packages/website/src/components/Playground.module.css]: this is some left over code from when i initially created it 👍 

thanks for catching this

```css
.codeContainer {
  display: flex;
  flex-direction: row;
  position: fixed;
  width: 100%;
  height: calc(100% - var(--ifm-navbar-height));
  top: var(--ifm-navbar-height);
}
```

**Comment by armano2** [comment]
there seem to be an issue with changing typescript version, I'm investigating this

**Comment by armano2** [comment]
i looked into your example and i devised something simmilar (havilly inspirated) implementation

### State and Editor

- state and editor has been refactored to not trigger update when its not needed
- console logs has been added
- data that comes from hash/editor/options is now correctly detected and only actions that are neccesary are executed
- responsiveness of editor should be improved as most of operations are not triggered on each operation, but rather we are waiting for user to stop typing

### Ast viewer:

- selection updating logic has been moved into viewer from playground
- if no node is selected / error occurs we should see correct error message

File: packages/website/src/components/modals/Modal.tsx
Comment (JoshuaKGoldberg) [reply packages/website/src/components/modals/Modal.tsx:13]: Ha, yeah... I'm down to hold off of them for now and treat using them as a followup issue.

File: packages/website/src/components/OptionsSelector.tsx
Comment (JoshuaKGoldberg) [inline packages/website/src/components/OptionsSelector.tsx:186]: Very non-blocking ideation: I wonder if we could use a schema/factory approach to automate creating this side... it'd probably be more code and probably not worth it 🤷

```ts
const sections = [
  {
    heading: "Info",
    inputs: [
      {
        label: "TypeScript",
        options: [ /* ... */ ]
        type: "select",
      },
      // ...
    ],
  },
  // ...
];
```

```ts
sections.map(section => <SidebarSection {...section} key={section.heading} />);
```

**Review by JoshuaKGoldberg — APPROVED** [review]
👏 Love it! Let's get this shipped!!

File: packages/website/src/components/OptionsSelector.tsx
Comment (bradzacher) [reply packages/website/src/components/OptionsSelector.tsx:186]: let's follow up on this with a future PR -  this is already v big.

**Comment by bradzacher** [comment]
banderas_yes

**Comment by bradzacher** [comment]
ugh the lint job OOM'd - we might need #4197 first?

**Comment by Josh-Cena** [comment]
This is... amazing. So well integrated with the rest of the website in terms of design and UX. Well done 👍 Do you guys plan to add this to the Docusaurus showcase?

Several nits though:

The type hint is obstructed by the navbar. (Not sure what to do here, adjust `z-index`?)

image

We need an ergonomic way to batch-enable rules, e.g. enabling recommended, enabling all, enabling all fixable rules... (I also hope rule options can be configured through the interface, but that may be too much work)

Maybe `⌘ + S` should copy the link to clipboard, like in the TS playground?

**Comment by armano2** [comment]
> We need an ergonomic way to batch-enable rules, e.g. enabling recommended, enabling all, enabling all fixable rules... (I also hope rule options can be configured through the interface, but that may be too much work)

https://github.com/orgs/typescript-eslint/projects/3/views/1 - its on todo list

**Comment by armano2** [comment]
small update, i noticed that there was small issue with ast viewer (there was html updates even without code change) 
that could cause high cpu usage now this issue is fixed, while on it i did some small design/ux improvements to it

image

**Comment by armano2** [comment]
> The type hint is obstructed by the navbar. (Not sure what to do here, adjust `z-index`?)
> 
> image

fixed

**Comment by bradzacher** [comment]
one feature request to add to the todo later list for this:

we should add a panel where you can view the "fix" output.
That'd be great as it'd let us repro fix breakages (like #4090) concretely.

File: packages/website-eslint/src/linter/CompilerHost.js
Comment (bradzacher) [inline packages/website-eslint/src/linter/CompilerHost.js:1]: question: could we leverage [`@typescript/vfs`](https://www.npmjs.com/package/@typescript/vfs) for some of this infra?

File: packages/website-eslint/src/linter/parser.js
Comment (bradzacher) [inline packages/website-eslint/src/linter/parser.js:38]: extraneous comment?
there's no code on the next line that gets ignored by this

File: packages/website-eslint/src/mock/assert.js
Comment (bradzacher) [inline packages/website-eslint/src/mock/assert.js:1]: is there a licence header we should embed in this file?

File: packages/website-eslint/src/mock/assert.js
Comment (bradzacher) [inline packages/website-eslint/src/mock/assert.js:2]: do we want to modernize this file a bit?
eg replace `var` with `const`?

File: packages/website-eslint/src/mock/path.js
Comment (bradzacher) [inline packages/website-eslint/src/mock/path.js:1]: should we modernize this file (eg replace `var` with `const`)?

**Review by bradzacher — COMMENTED** [review]
if you merge main in then it should fix the linter oom as well

File: packages/website-eslint/src/mock/assert.js
Comment (armano2) [reply packages/website-eslint/src/mock/assert.js:1]: hmm, i think we should just copy, same as i did in mock/path.js
https://github.com/nodejs/node/blob/master/LICENSE

File: packages/website-eslint/src/linter/CompilerHost.js
Comment (armano2) [reply packages/website-eslint/src/linter/CompilerHost.js:1]: we could, but i will prefer to do this after initial code is merged, 

generally I'd like to remove ts compilation from this code, and reuse AST from sandbox, but that's going to require more changes

File: packages/website-eslint/src/linter/parser.js
Comment (armano2) [reply packages/website-eslint/src/linter/parser.js:38]: i missed that when i extracted build from website, good catch

File: packages/website-eslint/rollup.config.js
Comment (armano2) [reply packages/website-eslint/rollup.config.js:82]: we should actually mess with this, this version acxtually simply removes files that should not be exposed to web

eg. currently if you try to import plugins you are importing every single file in every single project, and there is no treeshaking possible, even if you remove some of those stuff, this should still work as long as you polyfill more libs

by doing those overrides i went down from 30mb to 1-2mb of code (there is still alot of stuff that is not removed and it should)

most of issues comes from experimental utils and all of those `* as X` as its compiled to 
```
__importStar(require('name'))
```

witch makes code untreeshakable

File: packages/website-eslint/src/linter/CompilerHost.js
Comment (armano2) [reply packages/website-eslint/src/linter/CompilerHost.js:1]: as for now we are actually using `@typescript/vfs` but not in so obvious way

1. we are compiling lib with typescript as external library
2. we are using typescript as window.ts
3. windows.ts is provided by sandbox

**Comment by armano2** [comment]
> we should add a panel where you can view the "fix" output.
> That'd be great as it'd let us repro fix breakages (like #4090) concretely.

added to https://github.com/orgs/typescript-eslint/projects/3/views/1

File: packages/website-eslint/src/linter/CompilerHost.js
Comment (bradzacher) [reply packages/website-eslint/src/linter/CompilerHost.js:1]: Sure thing! Happy to leave this as a follow up.

**Comment by bradzacher** [comment]
We should talk to / learn from the TS folks with respect to their multi-file playground. I know they added it for their Halloween event and want to enable it more broadly. It'd be good for more complex repros down the line.

(Obv not part of this PR)

**Comment by bradzacher** [comment]
Fix up the lint errors and then we can merge this in 😃

**Comment by armano2** [comment]
> Fix up the lint errors and then we can merge this in 😃

fixed

File: packages/website/.eslintrc.js
Comment (armano2) [inline packages/website/.eslintrc.js:25]: for now, i'm disabling it, we have to deep dive to all of them

File: packages/website/.eslintrc.js
Comment (armano2) [inline packages/website/.eslintrc.js:24]: we do not want to use enums in here, we do not import `typescript-eslint` on web

**Comment by armano2** [comment]
not sure why nx is failing here on netlify, looks like its cache issue

**Comment by bradzacher** [comment]
https://github.com/armano2/typescript-eslint/blob/website-playground/workspace.json

I think you need to add the package here?

I don't know much about the setup tho. James is the expert with it

**Comment by armano2** [comment]
> https://github.com/armano2/typescript-eslint/blob/website-playground/workspace.json
> 
> I think you need to add the package here?
> 
> I don't know much about the setup tho. James is the expert with it

but `@typescript-eslint/website-eslint` is already there, hmm

**Comment by bradzacher** [comment]
```
9:37:24 PM: /opt/build/repo/node_modules/@nrwl/workspace/src/utilities/project-graph-utils.js:8

9:37:24 PM: return project.data && project.data.targets && project.data.targets[target];

9:37:24 PM: ^

9:37:24 PM: TypeError: Cannot read properties of undefined (reading 'data')

9:37:24 PM: at Object.projectHasTarget (/opt/build/repo/node_modules/@nrwl/workspace/src/utilities/project-graph-utils.js:8:20)

9:37:24 PM: at addTasksForProjectDependencyConfig (/opt/build/repo/node_modules/@nrwl/workspace/src/tasks-runner/run-command.js:168:43)

9:37:24 PM: at addTasksForProjectTarget (/opt/build/repo/node_modules/@nrwl/workspace/src/tasks-runner/run-command.js:106:13)

9:37:24 PM: at addTasksForProjectDependencyConfig (/opt/build/repo/node_modules/@nrwl/workspace/src/tasks-runner/run-command.js:169:21)

9:37:24 PM: at addTasksForProjectTarget (/opt/build/repo/node_modules/@nrwl/workspace/src/tasks-runner/run-command.js:106:13)

9:37:24 PM: at addTasksForProjectDependencyConfig (/opt/build/repo/node_modules/@nrwl/workspace/src/tasks-runner/run-command.js:169:21)

9:37:24 PM: at addTasksForProjectTarget (/opt/build/repo/node_modules/@nrwl/workspace/src/tasks-runner/run-command.js:106:13)

9:37:24 PM: at addTasksForProjectDependencyConfig (/opt/build/repo/node_modules/@nrwl/workspace/src/tasks-runner/run-command.js:169:21)

9:37:24 PM: at addTasksForProjectTarget (/opt/build/repo/node_modules/@nrwl/workspace/src/tasks-runner/run-command.js:106:13)

9:37:24 PM: at createTasksForProjectToRun (/opt/build/repo/node_modules/@nrwl/workspace/src/tasks-runner/run-command.js:90:9)


```

@JamesHenry might have an idea of what's causing the above crash?

**Comment by JamesHenry** [comment]
@bradzacher @armano2 Fixed!

I didn't bother to look into the specific error too deeply because it was occurring when building all projects as part of the postinstall process. That is repeated work for us because we already do that in the GitHub Actions and so I instead figure out how to skip the postinstall within the Netlify build.

I just had to set the `YARN_FLAGS` environment variable equal to `--ignore-scripts` and that did the trick.

Thanks so much again for your great work on this @armano2!

**Review by JamesHenry — APPROVED** [review]
🎉

**Comment by bradzacher** [comment]
I just used this to inspect the AST for TS 4.5 import/export specifier kinds so I could build scope analysis support.
It's truly awesome! astexplorer.net is so far behind (4.1.0 of our package - 1 year old) - so I couldn't use it - but our website is completely up to date!

Such a HUGE win here. I'm so happy you built this with AST support!

**Pull request merged**
